### PR TITLE
[Serve] Support serve component callback

### LINKED ISSUE #35046: [serve] Allow callback to be called in HTTPProxy constructor.
### Description

Allow custom callback to run in HTTPProxy actor.

### Use case

We would like to do some customization on all our deployments and actors. We can do that in the constructor, except for the HTTPProxy. example of things we would like to do

- logging config:
   - patch makeRecord: include our own request id from our own global context.
   - modify handlers: include the newly added info from makeRecord by changing the formatter of handlers.
- Connect to vscode debugger from their [API](https://github.com/microsoft/debugpy/blob/main/src/debugpy/public_api.py#L123)

### PR DESCRIPTION
**Pull request by sihanwang41**
## Why are these changes needed?

- User can set environment variable to setup controller and http proxy callback function path.

test_callback.py

```
from starlette.middleware import Middleware
import logging

logger = logging.getLogger("ray.serve")

class ASGIMiddleware:
    def __init__(self, app):
        self.app = app

    async def __call__(self, scope, receive, send):
        logger.info("my_logger!")
        await self.app(scope, receive, send)


def proxy_callback():
    return [Middleware(ASGIMiddleware)]
```
set env variable to start ray serve.
```
RAY_SERVE_HTTP_PROXY_CALLBACK_IMPORT_PATH="test_callback.proxy_callback"
```

@shayanhoshyari manual tested for their use case.

## Related issue number

Closes: #35046

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by shayanhoshyari** [comment]
I did not have much luck with this. I ran into some problems.

Here I fully put a completely standalone example https://github.com/shayanhoshyari/ray-playground/tree/main/http_proxy_callback with what I am trying to achieve and what is failing. Here I have the vscode workflow, but the logger and middlewares are also similar.

I tried the callback for init logging and [debugger workflow](https://github.com/ray-project/ray/issues/35674). See the comments in `main.py` and `readme.md`. The summary of the issue is that I don't have a way to pass the env var that contains the path to the callback, and the information that the callback needs to the HTTPProxy. 

I did not try the production workflow. My main requirement there is to be able to put the callback in runtime_env
   ```
   import_path: fruit:deployment_graph
   runtime_env: {
       # Can HTTP Proxy pick this up?
       "SERVE_HTTP_PROXY_CALLBACK_IMPORT_PATH" : "fruit:http_callback"
   }
   ```

**Comment by sihanwang41** [comment]
> I did not have much luck with this. I ran into some problems.
> 
> Here I fully put a completely standalone example https://github.com/shayanhoshyari/ray-playground/tree/main/http_proxy_callback with what I am trying to achieve and what is failing. Here I have the vscode workflow, but the logger and middlewares are also similar.
> 
> I tried the callback for init logging and [debugger workflow](https://github.com/ray-project/ray/issues/35674). See the comments in `main.py` and `readme.md`. The summary of the issue is that I don't have a way to pass the env var that contains the path to the callback, and the information that the callback needs to the HTTPProxy.
> 
> I did not try the production workflow. My main requirement there is to be able to put the callback in runtime_env; e.g.
> 
> ```
> import_path: fruit:deployment_graph
> runtime_env: {
>     # Can HTTP Proxy pick this up?
>     "py_modules" : "...", # code that contains fruit.
>     "SERVE_HTTP_PROXY_CALLBACK_IMPORT_PATH" : "fruit:http_callback"
> }
> ```

Hi @shayanhoshyari ,  this can be unblocked 
1. Set the environment in your image directly.
2. Set env_var into your ray process start command. (https://docs.ray.io/en/latest/serve/dev-workflow.html?highlight=env_vars#testing-on-a-remote-cluster)

File: python/ray/serve/controller.py [hunk 1]
```diff
@@ -179,6 +180,19 @@ async def __init__(
...
+                raise ValueError(
+                    f"Controller callback {RAY_SERVE_CONTROLLER_CALLBACK_IMPORT_PATH} "
+                    "is not callable."
+                )
+            logger.info(
+                "Calling controller callback func "
+                f"{RAY_SERVE_CONTROLLER_CALLBACK_IMPORT_PATH}"
+            )
+            controller_callback()
+
```
Comment (shrekris-anyscale) [inline python/ray/serve/controller.py:195]: Could you add a `logger.info()` line after the callback is finished?

**Review by shrekris-anyscale — COMMENTED** [review]
Nice work so far!

**Review by edoakes — COMMENTED** [review]
@sihanwang41 the logic for the test is pretty hard to follow because it's scattered across three places: `BUILD`, `tests/resources/callback.py`, and the test file.

You should be able to put this all in one file (`test_callback.py`) instead. Just have the import path be like `ray.serve.tests.test_callback.http_proxy_callback`. Then write a fixture that sets the relevant environment variables and then starts Serve.

File: python/ray/serve/controller.py [hunk 2]
```diff
@@ -179,6 +180,19 @@ async def __init__(
 
         self._recover_config_from_checkpoint()
 
+        if RAY_SERVE_CONTROLLER_CALLBACK_IMPORT_PATH:
```
Comment (edoakes) [inline python/ray/serve/controller.py:183]: wrap this in a method please and make sure errors are handled gracefully

File: python/ray/serve/tests/test_callback.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
+import pytest
+import requests
+import starlette
+import sys
+
+from ray import serve
+
+
+def test_callback(serve_instance, capsys):
```
Comment (edoakes) [inline python/ray/serve/tests/test_callback.py:9]: please add tests cases for each where:

1. the import paths are invalid
2. the callbacks raise an exception

Comment (edoakes) [reply python/ray/serve/tests/test_callback.py:9]: also for HTTP proxy, add a test case with a bad return type (not middlewares)

File: python/ray/serve/_private/http_proxy.py [hunk 4]
```diff
@@ -624,6 +626,25 @@ def __init__(
...
+                raise ValueError(
+                    f"HTTP proxy callback {RAY_SERVE_HTTP_PROXY_CALLBACK_IMPORT_PATH} "
+                    "is not callable."
+                )
+            logger.info(
+                "Calling HTTP proxy callback func "
+                f"{RAY_SERVE_HTTP_PROXY_CALLBACK_IMPORT_PATH}"
+            )
+            middlewares = proxy_callback()
+            if not isinstance(middlewares, list):
```
Comment (edoakes) [inline python/ray/serve/_private/http_proxy.py:642]: this should handle `None` as equivalent to `[]` IMO

File: python/ray/serve/tests/resources/callback.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
+from starlette.middleware import Middleware
```
Comment (edoakes) [inline python/ray/serve/tests/resources/callback.py:1]: I think you can delete this file now

File: python/ray/serve/tests/test_callback.py [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+import sys
+
+
+import ray
+from ray.exceptions import RayActorError
+from ray import serve
+from ray.serve._private.utils import call_function_from_import_path
+
+
+# ==== Callback function begin ====
```
Comment (edoakes) [inline python/ray/serve/tests/test_callback.py:17]:
```suggestion
# ==== Callbacks used in this test ====
```

File: python/ray/serve/tests/test_callback.py [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+    """Test invalid callback path in http proxy"""
+
+    actor_def = ray.serve._private.http_proxy.HTTPProxyActor
+    handle = actor_def.remote(
+        host="http_proxy",
+        port=123,
+        root_path="/",
+        controller_name="controller",
+        node_ip_address="127.0.0.1",
+    )
```
Comment (edoakes) [inline python/ray/serve/tests/test_callback.py:197]: can we test the actual user-facing behavior instead of this?

i.e., check that the proxy status goes to `UNHEALTHY` and that error message shows up in the controller log

Comment (edoakes) [reply python/ray/serve/tests/test_callback.py:197]: if checking controller log is hard/hacky that's OK

Comment (sihanwang41) [reply python/ray/serve/tests/test_callback.py:198]: It is doable, added a test check controller log and make sure http proxy is set to be unhealthy.

Comment (sihanwang41) [reply python/ray/serve/tests/test_callback.py:198]: (ideally we should have a separate test for http proxy, and make sure any exception from constructor can lead to http proxy state to be `UNHEALTHY`.)

File: python/ray/serve/_private/http_proxy.py [hunk 8]
```diff
@@ -615,6 +616,32 @@ def __init__(
...
+                for middleware in middlewares:
+                    if not issubclass(
+                        type(middleware), starlette.middleware.Middleware
+                    ):
+                        raise ValueError(
+                            "HTTP proxy callback "
+                            f"{RAY_SERVE_HTTP_PROXY_CALLBACK_IMPORT_PATH} "
+                            f"must return a list of Starlette middlewares, "
+                            f"instead got {type(middleware)} type item in the list."
+                        )
```
Comment (edoakes) [inline python/ray/serve/_private/http_proxy.py:642]: would be nice to put this in its own method/function to improve readability
```suggestion
            middlewares = _validate_callback_return(call_function_from_import_path(
                RAY_SERVE_HTTP_PROXY_CALLBACK_IMPORT_PATH
            ))
```

File: python/ray/serve/_private/utils.py [hunk 9]
```diff
@@ -638,3 +639,31 @@ def send_forever():
...
+        import_path: The import path of the function to call.
+    Raises:
+        ValueError: If the import path is invalid.
+        TypeError: If the import path is not callable.
+        RuntimeError: if the function raise exeception during execution.
+    Returns:
+        The result of the function call.
+    """
+    try:
+        proxy_callback = import_attr(import_path)
```
Comment (edoakes) [inline python/ray/serve/_private/utils.py:657]: "proxy" is an abstraction leak, please fix naming

File: python/ray/serve/controller.py [hunk 10]
```diff
@@ -113,6 +115,10 @@ async def __init__(
         configure_component_logger(
             component_name="controller", component_id=str(os.getpid())
         )
+        if RAY_SERVE_CONTROLLER_CALLBACK_IMPORT_PATH:
+            logger.info("controller_callback_called")
+            print("controller_callback_called")
```
Comment (edoakes) [inline python/ray/serve/controller.py:120]:
```suggestion
            logger.info(f"Calling user-provided callback from import path {RAY_SERVE_CONTROLLER_CALLBACK_IMPORT_PATH}")
```

**Review by edoakes — APPROVED** [review]
